ddl/partition.go

```diff
@@ -1206,3 +1208,56 @@ func truncateTableByReassignPartitionIDs(t *meta.Meta, tblInfo *model.TableInfo)
 	tblInfo.Partition.Definitions = newDefs
 	return nil
 }
+
+func checkPlacementRules(t *meta.Meta, job *model.Job, rules []*placement.Rule) ([]*placement.Rule, error) {
+	tblInfo, err := getTableInfo(t, job.TableID, job.SchemaID)
+	if err != nil {
+		return nil, err
+	}
+
+	ptInfo := tblInfo.GetPartitionInfo()
+	leftRules := rules[:0]
+
+	for _, rule := range rules {
+		pid, _ := strconv.ParseInt(rule.GroupID, 10, 64)
+		if ptInfo.GetNameByID(pid) != "" {
+			leftRules = append(leftRules, rule)
+		}
+	}
+
+	return leftRules, nil
+}
+
+func onAlterTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
+	var rules []*placement.Rule
+	err := job.DecodeArgs(&rules)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return 0, errors.Trace(err)
+	}
+
+	ver, err := t.GetSchemaVersion()
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	rules, err = checkPlacementRules(t, job, rules)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	ok, err := infosync.UpdatePlacementRules(nil, rules)
+	if err != nil {
+		if ok {
+			// ok means there is one HTTP resp, but not 200
+			// Cancel the job if PD declined us for some reason, avoiding meaningless retries
+			job.State = model.JobStateCancelled
+			err = errors.Wrapf(err, "PD refused the placement rules")
+		}
+		return 0, errors.Trace(err)
```

Comment:
I think it doesn't matter whether PD refuses it or PD is unavailable. It's all unexpected and the state should be `JobStateCancelled`.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/d086f02432e7bfaca0a16c8a3c282fc648f7d1bc